fragment/fragment/src/main/java/androidx/fragment/app/strictmode/FragmentStrictMode.java

```diff
@@ -0,0 +1,197 @@
+         * what we should do when we detect a problem.
+         *
+         * <p>You can call as many {@code detect} and {@code penalty} methods as you like. Currently
+         * order is insignificant: all penalties apply to all detected problems.
+         */
+        public static final class Builder {
+            private final Set<Flag> flags;
+            private OnViolationListener listener;
+
+            /** Create a Builder that detects nothing and has no violations. */
+            public Builder() {
+                flags = new HashSet<>();
+            }
+
+            /** Log detected violations to the system log. */
+            @NonNull
+            public Builder penaltyLog() {
+                flags.add(Flag.PENALTY_LOG);
+                return this;
+            }
+
+            /**
+             * Crash the whole process on violation. This penalty runs at the end of all enabled
+             * penalties so you'll still get to see logging or other violations before the process
+             * dies.
+             */
+            @NonNull
+            public Builder penaltyDeath() {
+                flags.add(Flag.PENALTY_DEATH);
+                return this;
+            }
+
+            /** Call #{@link OnViolationListener#onViolation} for every violation. */
+            @NonNull
+            public Builder penaltyListener(@NonNull OnViolationListener listener) {
+                this.listener = listener;
+                return this;
+            }
+
+            /**
+             * Construct the Policy instance.
+             *
+             * <p>Note: if no penalties are enabled before calling <code>build</code>, {@link
+             * #penaltyLog} is implicitly set.
+             */
+            @NonNull
+            public Policy build() {
+                if (listener == null && !flags.contains(Flag.PENALTY_DEATH)) {
+                    penaltyLog();
+                }
+                return new Policy(flags, listener);
+            }
+        }
+    }
+
+    /** Returns the current default policy. */
+     @NonNull
+    public static Policy getDefaultPolicy() {
+        return defaultPolicy;
+    }
+
+    /**
+     * Sets the policy for what actions should be detected, as well as the penalty if such actions
+     * occur.
+     *
+     * @param policy the policy to put into place
+     */
+    public static void setDefaultPolicy(@NonNull Policy policy) {
+        defaultPolicy = policy;
+    }
+
+    private static Policy getNearestPolicy(@Nullable Fragment fragment) {
+        while (fragment != null) {
+            if (fragment.isAdded()) {
+                FragmentManager fragmentManager = fragment.getParentFragmentManager();
+                if (fragmentManager.getStrictModePolicy() != null) {
+                    return fragmentManager.getStrictModePolicy();
+                }
+            }
+            fragment = fragment.getParentFragment();
+        }
+        return defaultPolicy;
+    }
+
```

Comment:
Sample hook for violations could look like this:
```java
public static void onDemoViolation(@NonNull Fragment fragment) {
    Policy policy = getNearestPolicy(fragment);
    if (policy.flags.contains(Flag.DETECT_DEMO)) {
        onPolicyViolation(policy, new DemoViolation());
    }
}
```
This would be called from wherever the violation is triggered.